src/spawn.c

```diff
@@ -0,0 +1,1121 @@
+ * will generally work. They open the child in a normal window, while spawn_...()
+ * use a minimized-by-default window, more suitable for a tool.
+ *
+ * This module does not depend on Geany.
+ */
+
+#ifdef HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#include <errno.h>
+#include <fcntl.h>
+#include <string.h>
+
+#ifdef G_OS_WIN32
+# define _NO_W32_PSEUDO_MODIFIERS
+# include <windows.h>
+# include <errno.h>
+# include <io.h>
+#else  /* G_OS_WIN32 */
+# include <sys/types.h>
+# include <sys/wait.h>
+# include <signal.h>
+# include <unistd.h>
+# ifndef O_NONBLOCK
+# define O_NONBLOCK O_NDELAY
+# endif
+#endif  /* G_OS_WIN32 */
+
+#include "spawn.h"
+
+#ifndef _
+# define _
+#endif
+
+#ifdef G_OS_WIN32
+# define BLANKS " \t"
+#endif
+
+
+/**
+ *  Checks whether a command line is valid.
+ *
+ *  All OS:
+ *     - any leading spaces and tabs are ignored
+ *     - a command consisting solely of spaces and tabs is invalid
+ *  *nix:
+ *     - the standard shell quoting and escaping rules are used, see @c g_shell_parse_argv()
+ *     - as a consequence, an unqouted # at the start of an argument comments to the end of line
+ *  Windows:
+ *     - a quoted program name must be entirely inside the quotes ("C:\Foo\Bar".pdf is actually
+ *       started as "C:\Foo\Bar.exe")
+ *     - the standard Windows quoting rules are used: double quote is escaped with slash, and
+ *       any literal slashes before a double quote must be duplicated.
+ *
+ *  @param command_line the command line to check.
+ *  @param execute whether to check (using @c g_find_program_in_path() if the program specified
+ *         in the command line exists and is executable.
+ *  @param error return location for error.
+ *
+ *  @return @c TRUE on success, @c FALSE if an error was set.
+ **/
+gboolean spawn_check_command(const gchar *command_line, gboolean execute, GError **error)
+{
+	gchar *program;
+
+#ifdef G_OS_WIN32
+	gboolean open_quote = FALSE;
+	const gchar *s, *arguments;
+
+	while (*command_line && strchr(BLANKS, *command_line))
+		command_line++;
+
+	if (!*command_line)
+	{
+		/* from glib */
+		g_set_error(error, G_SHELL_ERROR, G_SHELL_ERROR_EMPTY_STRING,
+			_("Text was empty (or contained only whitespace)"));
+		return FALSE;
+	}
+
+	if (*command_line == '"')
+	{
+		command_line++;
+		/* Windows allows "foo.exe, but we may have extra arguments */
+		if ((s = strchr(command_line, '"')) == NULL)
+		{
+			/* from glib, except for the fixed quote */
+			g_set_error(error, G_SHELL_ERROR, G_SHELL_ERROR_BAD_QUOTING,
+				_("Text ended before matching quote was found for %c."
+				  " (The text was '%s')"), '"', command_line);
+			return FALSE;
+		}
+
+		if (!strchr(BLANKS, s[1]))
+		{
+			g_set_error(error, G_SHELL_ERROR, G_SHELL_ERROR_BAD_QUOTING,
+				_("A quoted Windows program name must be entirely inside the quotes"));
+			return FALSE;
+		}
+	}
+	else
+	{
+		const gchar *quote = strchr(command_line, '"');
+
+		for (s = command_line; !strchr(BLANKS, *s); s++);

```

Comment:
depending on `strchr()` to find the `\0` is OK in practice, but I don't think it's obvious, so maybe adding an additional check to avoid confusion would be good.  Also, maybe having the `;` on the next, indented, line might make more obvious the body is empty.